### ISSUE TITLE
Mission failed not reporting mission type for some abandoned missions

### ISSUE THREAD

**Issue by Darkcyde13** [EDCD/EDDI#2247]
### Expected
When failing a mission, the mission type to be spoken.

### Observed
The mission type is not spoken in the Mission failed script, after the mission has already been abandoned.

### Steps to reproduce
1. Accept a mission (mainly megaship missions).
2. Abandon the mission.
3. Fail the mission (by clearing the abandoned mission from the in-game log).
4. Observe the Mission failed script does not say the mission type.

### Configuration
* Version [Please give full version numbers here, not just "latest" or "current"]
  * EDDI Version: v4.0.1
  * VoiceAttack Version: N/A
* Localization [if applicable]
  * EDDI's language: British English
  * Voice: Ivona 2 Amy British English

### Investigation Notes
I've finally got around to updating my personality for the new Missions scripts, and wanted to test a couple of things out that I had coded into my versions.  One was the disablemegaship mission type, and I wanted to see if I no longer needed my code now that tags have been implemented.

I took a disable mission, and the Mission accepted script spoke the mission type fine, so I was happy.  I then abandoned the mission, and the Mission abandoned script also spoke the mission type OK.  However, when I then cleared the abandoned mission from my log in-game, the Mission failed script fired as expected, but it did not speak the mission type.

I then tried the same method again for another megaship mission with the default personality, and the result was the same; the Mission failed script fails to report the mission type.

This is the output I got:

> You have accepted a Disable Megaship mission for Worker's Party of Hip 1 8 8 4 3.
> Disable Megaship mission for Worker's Party of Hip 1 8 8 4 3 abandoned.
> You have failed the mission.

> Hack Megaship mission for Party of Hip 1 8 8 4 3 accepted.
> Hack Megaship mission for Party of Hip 1 8 8 4 3 abandoned.
> mission failed.

Looking at the code for Mission failed, it would seem that mission.tags[0] is not populated.  So, I did one final check, and took another disable mission.  This time, I checked the EDDI mission monitor just after I abandoned the mission, and the log was empty.  So this appears to be why it's not working as expected; the mission is removed when abandoned, but is still needed for the Mission failed script.

I tried again with a normal salvage mission, but that did disappear from the in-game log when abandoned.  So my guess is that for disable megaship missions (and maybe others?) they should stay in the log until cleared manually in-game, which is the same as failing them.  I'm not sure how easy or hard that may be for you guys. :/

Alternatively, maybe check through the missions in the monitor for the failed mission ID, and if not found, just don't fire the Mission failed event script?  Or maybe do it in the script itself, something like this:

```
{for mission in missions:
    {if mission.missionid = event.missionid: {set found to true}}
}

{if found:
   (put script here)
}
```
I believe this will then not say anything if the mission is not found in the log to be 'failed'.  Well, it's just an idea. :)
Edit: I just tried this with another megaship mission, and it worked perfectly.  So this is a viable option if nothing else is better.
### EDDI Logs
N/A

### Player journals
These are the Journal entries:

```
{ "timestamp":"2021-08-29T22:37:38Z", "event":"MissionAccepted", "Faction":"Worker's Party of HIP 18843", "Name":"MISSION_DisableMegaship", "LocalisedName":"Disable the Megaship Turrets", "Target":"$MissionUtil_FactionTag_Turret;", "Target_Localised":"Turret", "Expiry":"2021-08-30T22:35:29Z", "Wing":false, "Influence":"++", "Reputation":"++", "Reward":1225416, "MissionID":802916489 }
{ "timestamp":"2021-08-29T22:49:38Z", "event":"MissionAbandoned", "Name":"MISSION_DisableMegaship_name", "MissionID":802916489 }
{ "timestamp":"2021-08-29T22:53:51Z", "event":"MissionFailed", "Name":"MISSION_DisableMegaship_name", "MissionID":802916489 }
```
```
{ "timestamp":"2021-08-29T23:03:22Z", "event":"MissionAccepted", "Faction":"Party of HIP 18843", "Name":"Mission_HackMegaship", "LocalisedName":"Infiltrate the Data Link Data Link", "Target":"$MissionUtil_FactionTag_Datalink;", "Target_Localised":"Data Link", "Wing":false, "Influence":"+", "Reputation":"+", "Reward":804670, "MissionID":802921033 }
{ "timestamp":"2021-08-29T23:03:45Z", "event":"MissionAbandoned", "Name":"Mission_HackMegaship_name", "MissionID":802921033 }
{ "timestamp":"2021-08-29T23:05:00Z", "event":"MissionFailed", "Name":"Mission_HackMegaship_name", "MissionID":802921033 }
```

**Comment by Tkael** [comment]
Hmm... you're right. We're removing the mission when it's abandoned which means that it's not retained / available for reference when the `Mission failed` event is triggered. We ought to instead update the mission status to `Failed` (or add a separate `Abandoned` state) and then remove it only after the `Mission failed` event.

https://github.com/EDCD/EDDI/blob/ff664ae5824e337c07de2b91587f6abbc1ed7a39/MissionMonitor/MissionMonitor.cs#L629-L651

**Comment by Darkcyde13** [comment]
I would agree, but it seems that for some missions, like the Salvage mission I tried above, that does disappear from the in-game log when abandoned, thus there is nothing left to 'fail'.

By leaving all 'Abandoned' missions in the Mission monitor, it would fill up as not all are then fail-able, meaning users will need to manually remove them all the time.  I know people probably don't go around abandoning missions all the time, but you get what I mean. :)

I can't see why FDev would leave some missions in the in-game log when they are abandoned, it seems pointless to me.  The only thing that can be done with them after abandonment, is to delete them, and that triggers a 'failed' entry in the Journal, even though technically it's not failed.

I would be in favour of still removing the mission from the Mission monitor when abandoned, but then have code to check if a failed mission is still there or not, and triggering the event only if it is.  So this would only trigger if a mission has genuinely failed normally, and not run the Mission failed event otherwise.  I guess it doesn't matter if that code is within EDDI itself, or just done in the script for Mission failed.  Although, being within EDDI would make it tidier and it doesn't really need to be exposed to the user in the script.

**Comment by Tkael** [comment]
The simplest solution is probably a `Mission failed` script change to make sure that the `mission` variable is set. Perhaps something like this?
```
{_ Context }
{SetState('eddi_context_last_subject', 'mission')}
{SetState('eddi_context_last_action', 'fail')}


{set mission to MissionDetails(event.missionid)}
{if mission:
    {if find(mission.invariantTags, "Transport") > -1:
        {set type to "Transport "}
        {set type to cat(type, mission.passengertype)}
    |else:
        {set type to mission.tags[0]}
    }
    {set type to cat(type, " mission")}

    {OneOf("{type} {if mission.faction: for {P(mission.faction, 'faction')}} failed",
        "You have failed the {type} {if mission.faction: for {P(mission.faction, 'faction')}}")}.
}
```

**Comment by Darkcyde13** [comment]
Ah yes, that looks perfect.  Sometimes the simplest solutions are the best! ;-)  I'm in favour of using your suggestion.

**Comment by Tkael** [comment]
Resolved by #a0f39a3af601e59f9924a8f872e162d354bc2bac.
